File: Two-Pointer/subarray-product-less-than-k.js
/**
 * 思路：
 * 求n個子陣列，或許可用sliding window來做，超出限制就shrink(move l++)，否則一直擴展window(move r++)
 * 當找到滿足條件的window時，去算該window下所有的子陣列排列組合有多少個。(r - l + 1)
 * 想得到公式，只需要去想在第i個元素時，答案會是多少，跑幾個例子就可以抓到這個pattern了
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numSubarrayProductLessThanK = function (nums, k) {
  if (k <= 1) {
    return 0;
  }

  let l = 0,
    ans = 0,
    cur = 1;

  for (let r = 0; r < nums.length; r++) {
    cur *= nums[r];

    while (cur >= k) {
      cur /= nums[l];
      l++;
    }

    ans += r - l + 1;
  }

  return ans;
};

File: Two-Pointer/subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        # 求x個subarrays,去想在每一個索引之下,滿足條件的子陣列有幾個,找看看有無數學規律
        # 此題規律： nums[i]有幾個子陣列 = 在這個索引的下的window長度 (r - l + 1) 
        # ex: [10,5,2,6] i = 2 window = 5(l = 1),2(r = 2) window長度=有幾個子陣列 2 - 1 + 1 = 2 
        if k <= 1:
            return 0

        curr = 1
        ans = 0
        l = 0

        for r in range(len(nums)):
            curr *= nums[r]

            while curr >= k:
                curr /= nums[l]
                l += 1
            
            ans += r - l + 1


        return ans